Lib/compression/zstd/_zstdfile.py

```diff
@@ -0,0 +1,378 @@
+
+        # File object
+        if isinstance(filename, (str, bytes, PathLike)):
+            if "b" not in mode:
+                mode += "b"
+            self._fp = builtins.open(filename, mode)
+            self._closefp = True
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
+        else:
+            raise TypeError("filename must be a str, bytes, file or PathLike object")
+        self._mode = mode_code
+
+        if self._mode == _MODE_READ:
+            raw = self._READER_CLASS(
+                self._fp,
+                ZstdDecompressor,
+                trailing_error=ZstdError,
+                zstd_dict=zstd_dict,
+                options=options,
+            )
+            self._buffer = io.BufferedReader(raw)
+
+    def close(self):
+        """Flush and close the file.
+
+        May be called more than once without error. Once the file is
+        closed, any other operation on it will raise a ValueError.
+        """
+        # Nop if already closed
+        if self._fp is None:
+            return
+        try:
+            if self._mode == _MODE_READ:
+                if hasattr(self, "_buffer") and self._buffer:
+                    self._buffer.close()
+                    self._buffer = None
+            elif self._mode == _MODE_WRITE:
+                self.flush(self.FLUSH_FRAME)
+                self._compressor = None
+        finally:
+            self._mode = _MODE_CLOSED
+            try:
+                if self._closefp:
+                    self._fp.close()
+            finally:
+                self._fp = None
+                self._closefp = False
+
+    def write(self, data):
+        """Write a bytes-like object to the file.
+
+        Returns the number of uncompressed bytes written, which is
+        always the length of data in bytes. Note that due to buffering,
+        the file on disk may not reflect the data written until .flush()
+        or .close() is called.
+        """
+        self._check_can_write()
+        if isinstance(data, (bytes, bytearray)):
+            length = len(data)
+        else:
+            # accept any data that supports the buffer protocol
+            data = memoryview(data)
+            length = data.nbytes
+
+        compressed = self._compressor.compress(data)
+        self._fp.write(compressed)
+        self._pos += length
+        return length
+
+    def flush(self, mode=FLUSH_BLOCK):
+        """Flush remaining data to the underlying stream.
+
+        The mode argument can be ZstdFile.FLUSH_BLOCK, ZstdFile.FLUSH_FRAME.
+        Abuse of this method will reduce compression ratio, use it only when
+        necessary.
+
+        If the program is interrupted afterwards, all data can be recovered.
+        To ensure saving to disk, also need to use os.fsync(fd).
+
+        This method does nothing in reading mode.
+        """
+        if self._mode == _MODE_READ:
+            return
```

Comment:
What do the other compression modules do on `flush()` to a file in reading mode? Wondering if it is better to raise an exception here.

Reply:
Most of them don't implement it! gzip does nothing in read mode: https://github.com/python/cpython/blob/2bbcaedb75942389dacb51866948f40de5951c9c/Lib/gzip.py#L406